# Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
# Define the path

path = r'/Users/iMac/Library/CloudStorage/OneDrive-Personal/Data Analytics/Data Analytics Immersion/Exercise 4/2024-01_Instacart Basket Analysis'

# Importing Datasets

In [3]:
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', '00 Prepared Data', 'ords_prods_merge.pkl'))

In [4]:
# create subset

df = ords_prods_merge[:1000000]

In [5]:
df.shape

(1000000, 14)

# If-Statements with User-Defined Functions

In [6]:
# Define price_label function

def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [7]:
# Apply price_label function

df['price_range'] = df.apply(price_label, axis = 1)

/var/folders/6c/bwpb08p57j1fmq58t70xfnf80000gq/T/ipykernel_88077/2831104905.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis = 1)


In [8]:
# Check 'price_range' frequency

df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    652638
Low-range product    338018
High range             9344
Name: count, dtype: int64

In [9]:
# check the most expensive product within the subset

df['prices'].max()

24.5

# If-Statements with the loc() Function

In [10]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

/var/folders/6c/bwpb08p57j1fmq58t70xfnf80000gq/T/ipykernel_88077/1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [11]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [12]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [13]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     652638
Low-range product     338018
High-range product      9344
Name: count, dtype: int64

## Apply the function on the entire df_ords_prods_merge dataframe instead of the df subset

In [14]:
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [15]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <= 15) & (ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [16]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [17]:
ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

# If-Statements with For-Loops

In [18]:
# Determine on which day most orders take place

ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [19]:
result = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [20]:
result

['Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Busiest day',
 'Least busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Bus

In [21]:
# Define 'busiest_day' column

ords_prods_merge['busiest_day'] = result

In [22]:
# Determine frequency of 'busiest_day' column

ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

# Task 4.7

## 2. Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [23]:
result = []

for value in ords_prods_merge["orders_day_of_week"]:
    if value in [0, 1]:
        result.append("Busiest days")
    elif value in [3, 4]:
        result.append("Least busy days")
    else:
        result.append("Regularly busy")

In [24]:
result

['Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Least busy days',
 'Least busy days',
 'Busiest days',
 'Regularly busy',
 'Least busy days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Least busy days',
 'Least busy days',
 'Regularly busy',
 'Least busy days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Least busy days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Least busy days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Least busy days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Least busy days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Least busy days',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',


In [25]:
# Define 'busiest_days' column

ords_prods_merge['busiest_days'] = result

In [26]:
ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 17 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   product_id              int64   
 1   product_name            object  
 2   aisle_id                int64   
 3   department_id           int64   
 4   prices                  float64 
 5   order_id                int64   
 6   user_id                 int64   
 7   order_number            int64   
 8   orders_day_of_week      int64   
 9   order_hour_of_day       int64   
 10  days_since_prior_order  float64 
 11  add_to_cart_order       int64   
 12  reordered               int64   
 13  _merge                  category
 14  price_range_loc         object  
 15  busiest_day             object  
 16  busiest_days            object  
dtypes: category(1), float64(2), int64(10), object(4)
memory usage: 3.9+ GB


## 3. Check the values of this new column for accuracy.

In [27]:
ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy     12916111
Busiest days       11864412
Least busy days     7624336
Name: count, dtype: int64

### Observations:

The majority of orders seem to fall into the "Regularly busy" category, with 12,916,111 occurrences.

The "Busiest days" category follows closely behind with 11,864,412 occurrences, indicating a substantial number of orders on the two busiest days.

The "Least busy days" category has 7,624,336 occurrences, suggesting fewer orders on the two least busy days.

## 4. When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [28]:
# Determine at which hours most orders take place

ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [29]:
result = []

for value in ords_prods_merge["order_hour_of_day"]:
    if value in [10, 11, 14, 15, 13, 12]:
        result.append("Most orders")
    elif value in [16, 9, 17, 8, 18, 19]:
        result.append("Fewest orders")
    else:
        result.append("Average orders")

In [30]:
result

['Most orders',
 'Fewest orders',
 'Average orders',
 'Most orders',
 'Fewest orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Fewest orders',
 'Fewest orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Fewest orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Fewest orders',
 'Fewest orders',
 'Most orders',
 'Fewest orders',
 'Fewest orders',
 'Most orders',
 'Most orders',
 'Fewest orders',
 'Fewest orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Fewest orders',
 'Fewest orders',
 'Few

In [31]:
# Define 'busiest_period_of_day' column

ords_prods_merge['busiest_period_of_day'] = result

## 5. Print the frequency for this new column.

In [32]:
ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       16128666
Fewest orders     11689984
Average orders     4586209
Name: count, dtype: int64

## 6. Ensure your notebook is clean and structured and that your code is well commented.

Check

## 7. Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder.

In [33]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','00 Prepared Data', 'orders_products_derivations.pkl'))